remote-read: streamed chunked server side; Extended protobuf; Added chunked, checksumed reader

bwplotka opened the pull request:
Part of: https://github.com/prometheus/prometheus/issues/4517 and https://github.com/improbable-eng/thanos/issues/488

Changes:
* Extended protobuf
* Chunked Writer/Reader.
* Server side implementation.

Intentionally there is:
* No optimization for older blocks against reencoding.
* Remote read client code

Those will be added in separate PR.

## Initial benchmarks are quite.. amazing: 
### Before:
* Avg request time: 25s (10s just Prometheus)
* Allocations vs Memory: 

![image](https://user-images.githubusercontent.com/6950331/59930666-3b46fc80-943b-11e9-8672-7ee4381cfd51.png)

### After:
* This really depends on configuration, but for now:
  * 120 samples in XOR chunk
  * Max 100 chunks per frame (probably too low).
* Avg request time 15s
* Frames:  160000
* Allocations vs Memory:

![image](https://user-images.githubusercontent.com/6950331/59930678-3eda8380-943b-11e9-99d4-97f9e7df3150.png)

CPU us 3x but we got here quite low chunk number, I am pretty sure we should try with bigger number.

More details on some decisions: https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit# 

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

brian-brazil commented on storage/remote/codec.go:
```diff
@@ -183,6 +183,148 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+
+	for isNext {
+		numSamples++
+		if sampleLimit > 0 && numSamples > sampleLimit {
```
Is it still possible to hit this?

bwplotka replied:
Yes, why not? Sample limit is counted overall for all frames, so `sampleLimit` here contains the left samples we can serve. 

With streaming it does not makes sense much maybe as we made remote read quite efficient with this, but still caller would like to limit number of overall samples maybe in all frames together? What do you think?

bwplotka replied:
The below error is wrong though as `sampleLimit` is not overall limit, just the left samples for the chunks here. Will fix the error message.

brian-brazil commented on storage/remote/codec.go:
```diff
@@ -183,6 +183,148 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+		chks       = make([]prompb.Chunk, 0, maxChunks)
+	)
+
+	for isNext {
```
I think you can do a more standard `for it := it.Next() {` here

bwplotka replied:
In current version I cannot due to:

> // encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).

This simplifies caller code for us, but happy to rearrange it a bit.

brian-brazil replied:
I'm not seeing that

bwplotka replied:
Hard to show you the issue, but this was the most clear implementation to me. Alternatives:

a)   passing `stream io.Write` into encodeChunks directly and stream straight away once chunks built up.

b)  duplicating `cut` process so it will end up like this:
```
// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, allowedNumSamples int) ([]prompb.Chunk, int, error) {
	var (
		chkMint int64
		chkMaxt int64
		chk     *chunkenc.XORChunk
		app     chunkenc.Appender
		err     error

		numSamples = 0
		chks       = make([]prompb.Chunk, 0, maxChunks)
	)

	for iter.Next() {
		numSamples++
		if allowedNumSamples > 0 && numSamples > allowedNumSamples {
			return nil, numSamples, HTTPError{
				msg:    fmt.Sprintf("exceeded sample limit"),
				status: http.StatusBadRequest,
			}
		}

		if chk == nil {
			chk = chunkenc.NewXORChunk()
			app, err = chk.Appender()
			if err != nil {
				return nil, numSamples, err
			}
			chkMint, _ = iter.At()
		}

		app.Append(iter.At())
		chkMaxt, _ = iter.At()

		if chk.NumSamples() < maxSamplesPerChunk {
			continue
		}

		// Cut the chunk.
		chks = append(chks, prompb.Chunk{
			MinTimeMs: chkMint,
			MaxTimeMs: chkMaxt,
			Type:      prompb.Chunk_Encoding(chk.Encoding()),
			Data:      chk.Bytes(),
		})
		chk = nil

		if maxChunks >= len(chks) {
			break
		}
	}
	if iter.Err() != nil {
		return nil, numSamples, errors.Wrap(iter.Err(), "iter TSDB series")
	}

	if chk != nil {
		// Cut the chunk if exists.
		chks = append(chks, prompb.Chunk{
			MinTimeMs: chkMint,
			MaxTimeMs: chkMaxt,
			Type:      prompb.Chunk_Encoding(chk.Encoding()),
			Data:      chk.Bytes(),
		})
	}
	return chks, numSamples, nil
}
```

bwplotka replied:
will try b let's see if we like it more

brian-brazil commented on prompb/remote.proto:
```diff
@@ -23,10 +23,32 @@ message WriteRequest {
+  // response_type allows negotiating content type of response.
+  // This does not guarantee the returned response. For servers that does not support this field, SAMPLES response type
+  // is used.
+  ResponseType response_type = 2;
```
Shouldn't this be a list, in order of preference?

bwplotka replied:
Very good question but will move this comment to https://github.com/prometheus/prometheus/pull/5701

brian-brazil commented on web/api/v1/api.go:
```diff
@@ -853,11 +854,18 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+	case prompb.ReadRequest_STREAMED_XOR_CHUNKS:
+		api.streamedChunkedRemoteRead(ctx, w, req)
+	default:
 		http.Error(w, fmt.Sprintf("%s response type is not implemented", req.ResponseType.String()), http.StatusNotImplemented)
```
This should fallback to SAMPLES

bwplotka replied:
hm.. Not sure about it, but if we move to `accepted_response_types` field instead, the logic will be clearer here, right? See https://github.com/prometheus/prometheus/pull/5701/files#r296731573

bwplotka replied:
Not sure, maybe for compatiblity falling back to sampled might make sense here

tomwilkie commented on cmd/prometheus/main.go:
```diff
@@ -212,6 +212,9 @@ func main() {
 
+	a.Flag("storage.remote.read-max-chunks-in-frame", "Maximum number of chunks in single frame for STREAMED_XOR_CHUNKS remote read response type. Each chunk corresponds roughly to (~3B * 120 samples) + 32B. Default is 1000 which is roughly (1000 * ~400B) + labelset, so approx. 0.4MB per frame .").
+		Default("1000").IntVar(&cfg.web.RemoteReadMaxChunksInFrame)
+
```
What happens when a single series has more than chunks than this?

tomwilkie replied:
I actually don't see where this is enforced...

bwplotka replied:
wait. I think there is misunderstanding. This is not a limit. It's essentially control on how much you want to buffer / how large frames are.

You can have many frames per series. How we can make it more clear here?

brian-brazil replied:
How about specifying this in bytes, so it works across future types?

tomwilkie replied:
@bwplotka yup, I misunderstood.  I see how this works now.  Thanks.

bwplotka replied:
Should we do it by bytes then? `storage.remote.read-max-bytes-in-frame` 

I think we might be able to calculate on the stream side, but it might be bit brittle and complex to implement.

brian-brazil replied:
We'd only need something that's good enough. The idea is more to stop massive frames, than to get it within the byte.

brian-brazil replied:
I still think we should either do this in bytes, or just leave it hardcoded.

tomwilkie commented on storage/remote/codec.go:
```diff
@@ -183,6 +183,150 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+}
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
+func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, allowedNumSamples int) ([]prompb.Chunk, int, error) {
```
If I read this correctly, this is iterating over the samples in the series and rebuilding the chunks.  Can we extend TSDB to just give us the chunks and save the hassle?

tomwilkie replied:
ps I want this for Cortex too, so I did it: https://github.com/cortexproject/cortex/pull/1427/files#diff-a6d94f56da1ed83021ef693d2dd12b55R58

brian-brazil replied:
I gather this PR is just the first step.

bwplotka replied:
Yes. I would really really vote for avoiding scope creep and merge working bits by bits.

We can add avoid re-encoding easily in other PRs, it's just internal optimization. I promise I will do it, but let's ship massive win with even this implementation. ;p

tomwilkie replied:
Happy to do this optimisation as another PR.  Even happy to do that myself.

bwplotka replied:
to link: https://github.com/prometheus/tsdb/pull/659

But I think iterator would be awesome instead.

tomwilkie commented on storage/remote/codec.go:
```diff
@@ -183,6 +183,150 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+				status: http.StatusBadRequest,
+			}
+		}
+
```
Do we need a sample limit in the streaming case?

bwplotka replied:
Good question, what are your thoughts? (: If you accidently ask for 10 years data you will just.. timeout I guess? so timeout might be totally enough.

bwplotka replied:
Happy to remove sample limitation

bwplotka replied:
Removed, and added info in sample limit flag.

tomwilkie commented on storage/remote/stream.go:
```diff
@@ -0,0 +1,99 @@
+		r.data = r.data[:size]
+	}
+
+	if _, err := io.ReadFull(r.b, r.data); err != nil {
```
I feel like we probably want to limit the size of this read (or the allocation above).

bwplotka replied:
Comment related to: https://github.com/prometheus/prometheus/pull/5702

bwplotka replied:
Will think about it, thanks!

tomwilkie commented on web/api/v1/api.go:
```diff
@@ -927,6 +963,85 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+			} else {
+				filteredMatchers = append(filteredMatchers, m)
+			}
+		}
```
Can we factor this code out and share it with the origin remote read implementation?

bwplotka replied:
I refactored a bit. I could refactor more (literally only response is different), but since it might change soon (samples vs fetching encoded chunks) I would not try to simplify it more. Let me know your thoughts.

tomwilkie replied:
We've had bugs in the selectParams not being honored in the past; super keen for this in particular not to be repeated.

bwplotka replied:
done

tomwilkie commented:
Looking good @bwplotka!  I've left any major points but not really gone through it with a fine tooth comb.

Personal preference for me would be to merge the first PRs into this one and not go this as a series of small PRs.  But thats up to you...

brian-brazil commented on prompb/remote.proto:
```diff
@@ -23,10 +23,32 @@ message WriteRequest {
+  // For request that do not support `accepted_response_types` field the non streamed, raw samples response is used.
+  // In such case response headers are as follows:
+  //
+  // Content-Type: "application/x-protobuf"
```
This should be the first item in the enum.

bwplotka replied:
See: https://github.com/prometheus/prometheus/pull/5701#discussion_r297801885

bwplotka replied:
Essentially the @tomwilkie idea was that SAMPLED is by default when the `accepted...` is empty, so no point in having separate enum item - and I can agree with that. It avoids having multiple options to control one thing.

What are you thoughts @brian-brazil why it should be in enum?

brian-brazil replied:
What if I explicitly want to ask for it as a fallback?

tomwilkie replied:
You can't.  It is the always the explicit fallback.  

If you want to ask for it explicitly, don't send any `accepted_response_types` - which is the same as getting a request for a client that doesn't know about this new field.

brian-brazil replied:
That's not great in negotiation terms, maybe it's too inefficient for what I'm trying to do. I'd rather keep this explicit.

tomwilkie replied:
Why is that not great?  Seems (a) pretty straight forward and (b) the correct embedding of what we want in what protos provide.

brian-brazil replied:
We'd send a whole pile of data that the other end doesn't want. I can see this as a gotcha in the future, why is the server returning something in a format I didn't ask for?

bwplotka replied:
Yea I can see the failure in case of unsupported accepted formats only as the more explicit logic. If the client wants to be backward compatible it might need to explicitly state some `SAMPLES` enum at the end of `accepted_response_type` and this is fair. 

On the other hand @tomwilkie is right that there will be now two ways to ask for SAMPLES flow: either empty `accepted_response_type` or explicit SAMPLES in `accepted_response_type`. I feel like this is less evil though than the bit of the confusion as well as lack of way to disable SAMPLES response totally from client side. And I think it can be fair use case.

I am happy to revert to previous flow with `SAMPLES` and `STREAMED_XOR_CHUNKS` enum once we have consenus on this. Any other thoughts? @gouthamve ?

brian-brazil commented on prompb/types.proto:
```diff
@@ -56,3 +57,25 @@ message ReadHints {
   int64 end_ms = 4;   // End time in milliseconds.
 }
+
+// Chunk represents a TSDB chunk.
```
open/closed should be documented here

bwplotka replied:
I guess you mean something like `// Time range [min, max] is inclusive.`?

bwplotka commented:
I might think that chained PRs in this form might be more confusion then benefit ;p

bwplotka commented:
OK, moved everything to this PR for now.

Updated description with:
```
3 commits:
* Extended protobuf
* Chunked Writer/Reader.
* Server side implementation.

Intentionally there is:
* No optimization for older blocks against reencoding.
* Remote read client code

Those will be added in separate PR.
```

bwplotka commented:
Need to update https://github.com/opentracing-contrib/go-stdlib dependency. In older version it does not support Flusher.

bwplotka commented on storage/remote/codec.go:
```diff
@@ -183,6 +164,154 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+		iter := series.Iterator()
+		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+
+		// TODO(bwplotka): We send each series in separate frame no matter what. Even if series has only one sample.
```
@brian-brazil @tomwilkie I think this is quite major optimization we can make, but adds extra complexity to the caller. I think it's worth it though.

brian-brazil replied:
I'm not sure it'll make much difference, series ordering should be preserved though.

tomwilkie replied:
Yeah this is fine to start, @bwplotka.  Could rephrase for clarity "A most one series per frame; series may be split over multiple frames according to maxChunksInFrame"

bwplotka commented:
Current state with 1000 max chunks:
![image](https://user-images.githubusercontent.com/6950331/60721068-a48f3b00-9f24-11e9-9ad8-3f31e0342229.png)

I think we could put *more* chunks into frame (2000)? and add optimization mentioned here: https://github.com/prometheus/prometheus/pull/5703#discussion_r300649511

bwplotka commented:
Prometheus mem spike is due to reencoding which is planned to be improved in later PRs.

bwplotka commented:
Build `bplotka/prometheus:v2.11.0-rc.0-rr-streaming` with remote read streaming server support (#5703) if someone want to test it.

bwplotka commented on storage/remote/codec.go:
```diff
@@ -183,6 +164,154 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
+func encodeChunks(iter storage.SeriesIterator, chks []prompb.Chunk, maxChunks int) ([]prompb.Chunk, error) {
+	const maxSamplesInChunk = 120
```
worth to comment that this hardcoded magic value is for XOR only

bwplotka replied:
actually we explicitly do XOR chunks here, so it's not needed.

bwplotka commented:
Not much changes indeed for `Series interface`  to return Chunks but quite major interface change: https://github.com/cortexproject/cortex/pull/1427/files#r301158203 Can experiment with this a bit a a separate side thing. Still looking for review on this PR if you anyone has some time.

brian-brazil commented on cmd/prometheus/main.go:
```diff
@@ -213,12 +213,15 @@ func main() {
 		Default("1m").PlaceHolder("<duration>").SetValue(&cfg.RemoteFlushDeadline)
 
-	a.Flag("storage.remote.read-sample-limit", "Maximum overall number of samples to return via the remote read interface, in a single query. 0 means no limit.").
+	a.Flag("storage.remote.read-sample-limit", "Maximum overall number of samples to return via the remote read interface, in a single query. 0 means no limit. This limit is ignored for STREAMED_XOR_CHUNKS response type.").
```
I'd say "streamed" to be future proof.

bwplotka replied:
good point!

brian-brazil commented on storage/remote/chunked.go:
```diff
@@ -0,0 +1,103 @@
+	return &ChunkedWriter{writer: w, flusher: f}
+}
+
+// Write writes given bytes to the stream. It adds uvarint delimiter before each message.
```
Should we add a checksum while we're at it?

tomwilkie replied:
I don't have strong opinions here; it adds complexity, OTOH TCP checksums are pretty useless.  I'm inclined to say not right now, mainly as we have the ability to introduce a new framing protocol with checksums in the future.

bwplotka replied:
isn't that done on [TCP level](https://locklessinc.com/articles/tcp_checksum/)?

brian-brazil replied:
As Tom says, TCP checksums are pretty useless.

tomwilkie commented on prompb/types.proto:
```diff
@@ -56,3 +57,25 @@ message ReadHints {
+message ChunkedSeries {
+  // labels should be sorted.
+  repeated Label labels = 1 [(gogoproto.nullable) = false];
+  // chunks should be sorted should not overlap in time.
```
Does this work with backfill? 

Also, Sentence case on comments please.

brian-brazil replied:
That's a case where we'd have to reencode I suspect.

tomwilkie replied:
Why?  

We could relax this constraint to something like "chunks will be in start time order and may overlap." and leave merging to the client.

brian-brazil replied:
That's extra complexity for every client to handle, and they'll likely only do so when the bug reports come in.

tomwilkie commented on storage/remote/chunked.go:
```diff
@@ -0,0 +1,103 @@
+		r.data = make([]byte, size)
+	} else {
+		r.data = r.data[:size]
+	}
```
We really need to have a limit to the reasonable lengths here; reading a corrupt length (perhaps due to a malicious server) could really hurt here.

bwplotka replied:
:+1:

bwplotka replied:
done.

tomwilkie commented on web/api/v1/api.go:
```diff
@@ -837,8 +840,23 @@ func (api *API) serveFlags(r *http.Request) apiFuncResult {
+func negotiateResponseType(accepted []prompb.ReadRequest_ResponseType) prompb.ReadRequest_ResponseType {
+	supported := map[prompb.ReadRequest_ResponseType]struct{}{
+		prompb.ReadRequest_STREAMED_XOR_CHUNKS: {},
+	}
```
Could probably be moved to the package level.

tomwilkie commented on web/api/v1/api.go:
```diff
@@ -853,39 +871,71 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+			if err := querier.Close(); err != nil {
+				level.Warn(api.logger).Log("msg", "error on querier close", "err", err.Error())
+			}
+		}()
```
This defer won't be executed until the whole function exits IIUC, meaning we can have multiple queries open per remote_read.  Might want to break the body of the loop out into its own function.

bwplotka replied:
:+1:

bwplotka replied:
done

tomwilkie commented on go.mod:
```diff
@@ -60,7 +60,7 @@ require (
 	github.com/oklog/run v1.0.0
 	github.com/olekukonko/tablewriter v0.0.1 // indirect
-	github.com/opentracing-contrib/go-stdlib v0.0.0-20170113013457-1de4cc2120e7
+	github.com/opentracing-contrib/go-stdlib v0.0.0-20190519235532-cf7a6c988dc9
```
Is this change spurious?

tomwilkie replied:
> Need to update https://github.com/opentracing-contrib/go-stdlib dependency. In older version it does not support Flusher.

bwplotka replied:
(:

tomwilkie commented on web/api/v1/api.go:
```diff
@@ -927,6 +977,82 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+		}
+		sort.Slice(sortedExternalLabels, func(i, j int) bool {
+			return sortedExternalLabels[i].Name < sortedExternalLabels[j].Name
+		})
```
Cloud this be done once, and not for each request?

tomwilkie commented on web/api/v1/api.go:
```diff
@@ -927,6 +977,82 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+		if !ok {
+			http.Error(w, "internal http.ResponseWriter does not implement http.Flusher interface", http.StatusInternalServerError)
+			return
+		}
```
When is this the case?

bwplotka replied:
When someone refactors other part of the code (wrap it with something that does not implement this).

We have lots of middlewares so it is super likely (: With this we can catch that in unit tests.

tomwilkie commented:
@bwplotka looking good.  Give me a ping when you've done another round of review comments, I'm hoping we can get this in for the next release.

bwplotka commented:
Addressed all @tomwilkie @brian-brazil (: Thanks for detailed reviewed so far! Ready for another one.

Still we need to agree on:
* Do we need checksum. I think we don't. (https://github.com/prometheus/prometheus/pull/5703#discussion_r302631617) 
* Should we have `SAMPLES` in ReadResponse type. I think i agree with @brian-brazil that we should. ( https://github.com/prometheus/prometheus/pull/5703#discussion_r304169343)

brian-brazil commented:
I'm not exactly awake right now, but it looks like just those to decide on.

bwplotka commented:
@tomwilkie do you have strong opinion about this? (:

>  Should we have SAMPLES in ReadResponse type. I think i agree with @brian-brazil that we should. (https://github.com/prometheus/prometheus/pull/5703#discussion_r304169343)

For the checksum I happy to add it straightaway.

bwplotka commented:
@tomwilkie @codesome @brian-brazil addressed all comments. 

* Added CRC32 checksum, plus tests.
* Reverted to explicit SAMPLES response type.
* Bit reluctant to change frame limit to be in bytes - I think having this in number of chunks is fair.

brian-brazil commented on prompb/remote.proto:
```diff
@@ -23,10 +23,35 @@ message WriteRequest {
+    // Response headers:
+    // Content-Type: "application/x-protobuf"
+    // Content-Encoding: "snappy"
+    SAMPLED = 0;
```
SAMPLES

brian-brazil commented on storage/remote/chunked.go:
```diff
@@ -0,0 +1,154 @@
+// Each frame includes:
+//
+// 1. uvarint for the size of the data frame.
+// 2. uvarint for the Castagnoli polynomial CRC-32 checksum of the data frame.
```
Why not fixed size?

bwplotka replied:
You are right, it should be fixed - CRC will be always fully filled

brian-brazil commented on storage/remote/codec.go:
```diff
@@ -183,6 +161,173 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
+	return nil
+}
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
```
s/done/called/

brian-brazil commented on web/api/v1/api.go:
```diff
@@ -856,78 +860,153 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
-		Results: make([]*prompb.QueryResult, len(req.Queries)),
+	externalLabels := api.config().GlobalConfig.ExternalLabels.Map()
+
+	// Add external labels back in, in sorted order.
```
This comment isn't right

bwplotka commented:
@brian-brazil in terms of limit of bytes per frame. How accurate you would be?

Sum:
* how many bytes labelset has
* add chunks one by one until bytes limit?

That number would probably will be near to the actual memory used or actual protobuf size neither ):

brian-brazil commented:
That's roughly what I'd do.

bwplotka commented:
Addressed comments and moved to bytes limt @brian-brazil

bwplotka commented:
Ping @brian-brazil @tomwilkie @codesome @krasi-georgiev if you have time. (:

brian-brazil commented on cmd/prometheus/main.go:
```diff
@@ -219,8 +219,8 @@ func main() {
-	a.Flag("storage.remote.read-max-chunks-in-frame", "Maximum number of chunks in single frame for STREAMED_XOR_CHUNKS remote read response type. Each chunk corresponds roughly to (~3B * 120 samples) + 32B. Default is 1000 which is roughly (1000 * ~400B) + labelset, so approx. 0.4MB per frame. Be aware that client might have limit on frame size as well.").
-		Default("1000").IntVar(&cfg.web.RemoteReadMaxChunksInFrame)
+	a.Flag("storage.remote.read-max-bytes-in-frame", "Maximum number of bytes in a single frame for streaming remote read response types before marshalling. Note that client might have limit on frame size as well.").
+		Default("1000").IntVar(&cfg.web.RemoteReadBytesInFrame)
```
This default needs adjusting.

bwplotka replied:
Yup. Moved to 1MB as probuf suggests that anything over 1MB should be streamed. Will perform some benchmarks with it.

brian-brazil commented on prompb/remote.proto:
```diff
@@ -23,10 +23,36 @@ message WriteRequest {
+  //
+  // Response types are taken from the list in the FIFO order. If no response type in `accepted_response_types` is
+  // implemented by server, error is returned.
+  // For request that do not support `accepted_response_types` field the SAMPLES response type will be used.
```
s/support/contain/

bwplotka replied:
done (:

bwplotka commented:
PTAL @brian-brazil @tomwilkie @krasi-georgiev 

I feel like we are close to merge this.

I will provide some benchmarks for this version. Let's merge and then apply `chunkIterator` once merged.

bwplotka commented:
Basic benchmark:

* 3 sequential requests:
* Used Images:
  * quay.io/thanos/prometheus:v2.12.0-rc.0-rr-streaming (built from this PR https://github.com/prometheus/prometheus/pull/5703/commits/f0af4045650963b02f75af9048e4787a2da59ecc)
  * quay.io/thanos/thanos:v0.6.0-master-rr-streaming (built from https://github.com/thanos-io/thanos/pull/1268)
* Max 1MB in frame.
* Avg request time 15s
* Frames:  10000
* Series: 1000
* Resources:

![2019-08-18-14:17:33-Selection](https://user-images.githubusercontent.com/6950331/63225079-8fb0f300-c1c4-11e9-9e97-0418b4308972.png)

It looks good, and from my side, this PR is ready to go. I plan to make more benchmarks with spam or requests etc while adjusting bytes number vs request concurrency limits, but all of this is in flags, and we can tweak defaults later on.

We will be rolling this on our clusters as well next week.

bwplotka commented on storage/remote/chunked.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+// NewChunkedReader constructs a ChunkedReader.
+func NewChunkedReader(r io.Reader, sizeLimit uint64) *ChunkedReader {
```
I think I will allow passing  `data []byte` as argument to this constructor to reuse byte slice even more.

krasi-georgiev commented on web/api/v1/api_test.go:
```diff
@@ -1012,10 +1013,188 @@ func TestReadEndpoint(t *testing.T) {
+	suite, err := promql.NewTest(t, `
+		load 1m
+			test_metric1{foo="bar1",baz="qux"} 0+100x119
+            test_metric1{foo="bar2",baz="qux"} 0+100x120
```
Comment says `// Second with 121 samples, We expect 1 frame with 2 chunks.` but here we have 120 why?

```suggestion
            test_metric1{foo="bar2",baz="qux"} 0+100x121
```

bwplotka replied:
so `0 + X` means X+1 samples (: 

Fixed `119 samples` comment as there are 120 exactly.

krasi-georgiev replied:
yep that was the confusion, just the comment than, thanks

krasi-georgiev commented on web/api/v1/api_test.go:
```diff
@@ -1012,10 +1013,188 @@ func TestReadEndpoint(t *testing.T) {
+			return config.Config{
+				GlobalConfig: config.GlobalConfig{
+					ExternalLabels: labels.Labels{
+						{Name: "baz", Value: "a"},
```
why are you adding this external label when it is already part of `test_metric1{foo="bar3",baz="qux"} 0+100x240` ?

bwplotka replied:
to check if external labels are handled properly

krasi-georgiev replied:
properly here means that adding an external label is ignored in this case?
Is this documented somewhere so I read about it? 
Maybe add a comment here as well.

bwplotka replied:
Added comment

krasi-georgiev commented on web/api/v1/api_test.go:
```diff
@@ -1012,10 +1013,188 @@ func TestReadEndpoint(t *testing.T) {
+							Type:      prompb.Chunk_XOR,
+							MaxTimeMs: 7140000,
+							Data:      []byte("\000x\000\000\000\000\000\000\000\000\000\340\324\003\302|\005\224\000\301\254}\351z2\320O\355\264n[\007\316\224\243md\371\320\375\032Pm\nS\235\016Q\255\006P\275\250\277\312\201Z\003(3\240R\207\332\005(\017\240\322\201\332=(\023\2402\203Z\007(w\2402\201Z\017(\023\265\227\364P\033@\245\007\364\nP\033C\245\002t\036P+@e\036\364\016Pk@e\002t:P;A\245\001\364\nS\373@\245\006t\ ...
+						},
```
Is there no way to get the expected chunk data from somewhere instead of hard coding it?
I checked briefly , but can't see any way to get the raw chunk bytes, maybe after adding the Chunk iterator?

I am thinking that if the implementation of the test suite changes at some point this would require refactoring the test as well.

bwplotka replied:
I know what you mean, but there is not much we can do here. Maybe, later on, we could potentially have a single place for proper encoding of chunks which will be tested in isolation and then use this here? For now, I would vote for leaving it hardcoded it's not that problematic.

krasi-georgiev replied:
When we merge the Chunk iterator should solve this, right?

bwplotka replied:
Potentially

brian-brazil commented on prompb/remote.proto:
```diff
@@ -43,3 +69,14 @@ message QueryResult {
+message ChunkedReadResponse {
+  repeated prometheus.ChunkedSeries chunked_series = 1;
+
+  // query_index represents an index of the query from ReadRequest.queries this results relates to.
```
s/this results/these chunks/

bwplotka commented on web/api/v1/api.go:
```diff
@@ -856,78 +860,150 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
+
+		f, ok := w.(http.Flusher)
+		if !ok {
+			http.Error(w, "internal http.ResponseWriter does not implement http.Flusher interface", http.StatusInternalServerError)
```
Isn't that a hard fail? Why panicking whole binary?

bwplotka replied:
We have test against that: https://github.com/prometheus/prometheus/blob/939d1b4263c26cc8256969a77de5948db01a53ca/web/api/v1/api_test.go#L1079

krasi-georgiev replied:
true, true!

krasi-georgiev commented:
:+1:

tomwilkie commented:
LGTM
